frame/staking/src/lib.rs

```diff
@@ -243,62 +243,98 @@
-//! is stored in the Session module's `Validators` at the end of each era.
+//! - [Session](../pallet_session/index.html): Used to manage sessions. Also, a list of new
+//!   validators is stored in the Session module's `Validators` at the end of each era.
 
-#![recursion_limit="128"]
+#![recursion_limit = "128"]
 #![cfg_attr(not(feature = "std"), no_std)]
 
 #[cfg(test)]
 mod mock;
 #[cfg(test)]
 mod tests;
-mod slashing;
+#[cfg(feature = "testing-utils")]
+pub mod testing_utils;
 #[cfg(any(feature = "runtime-benchmarks", test))]
 pub mod benchmarking;
 
+pub mod slashing;
+pub mod offchain_election;
 pub mod inflation;
 
-use sp_std::{prelude::*, result, collections::btree_map::BTreeMap};
+use sp_std::{
+	result,
+	prelude::*,
+	collections::btree_map::BTreeMap,
+	convert::{TryInto, From},
+	mem::size_of,
+};
 use codec::{HasCompact, Encode, Decode};
 use frame_support::{
-	decl_module, decl_event, decl_storage, ensure, decl_error, weights::SimpleDispatchInfo,
-	dispatch::DispatchResult, storage::IterableStorageMap, traits::{
+	decl_module, decl_event, decl_storage, ensure, decl_error, debug, Parameter, IterableStorageMap,
+	weights::SimpleDispatchInfo,
+	dispatch::{IsSubType, DispatchResult},
+	traits::{
 		Currency, LockIdentifier, LockableCurrency, WithdrawReasons, OnUnbalanced, Imbalance, Get,
-		Time
+		Time, EstimateNextNewSession, MigrateAccount,
 	}
 };
-use pallet_session::historical::SessionManager;
+use pallet_session::historical;
 use sp_runtime::{
-	Perbill, PerThing, RuntimeDebug,
+	Perbill, PerU16, PerThing, RuntimeDebug, RuntimeAppPublic,
 	curve::PiecewiseLinear,
 	traits::{
-		Convert, Zero, StaticLookup, CheckedSub, Saturating, SaturatedConversion,
-		AtLeast32Bit, EnsureOrigin,
-	}
+		Convert, Zero, StaticLookup, CheckedSub, Saturating, SaturatedConversion, AtLeast32Bit,
+		EnsureOrigin, Member, SignedExtension,
+	},
+	transaction_validity::{
+		TransactionValidityError, TransactionValidity, ValidTransaction, InvalidTransaction,
+		UnknownTransaction,
+	},
 };
 use sp_staking::{
 	SessionIndex,
 	offence::{OnOffenceHandler, OffenceDetails, Offence, ReportOffence, OffenceError},
 };
 #[cfg(feature = "std")]
 use sp_runtime::{Serialize, Deserialize};
-use frame_system::{self as system, ensure_signed, ensure_root};
-
-use sp_phragmen::ExtendedBalance;
-use frame_support::traits::MigrateAccount;
+use frame_system::{
+	self as system, ensure_signed, ensure_root, ensure_none,
+	offchain::SubmitUnsignedTransaction,
+};
+use sp_phragmen::{
+	ExtendedBalance, Assignment, PhragmenScore, PhragmenResult, build_support_map, evaluate_support,
+	elect, generate_compact_solution_type, is_score_better, VotingLimit, SupportMap,
+};
 
 const DEFAULT_MINIMUM_VALIDATOR_COUNT: u32 = 4;
-pub const MAX_NOMINATIONS: usize = 16;
-const MAX_UNLOCKING_CHUNKS: usize = 32;
 const STAKING_ID: LockIdentifier = *b"staking ";
+pub const MAX_UNLOCKING_CHUNKS: usize = 32;
+pub const MAX_NOMINATIONS: usize = <CompactAssignments as VotingLimit>::LIMIT;
```

Comment:
Does it make sense to move all of these to configuration traits?

Maybe not in your PR, but in general.

Reply:
Yes for the future.